### DISCUSSION TITLE
How to transfrom object props into one

### DISCUSSION THREAD

**Discussion by stychu** [colinhacks/zod#2177] [Q&A] [answered]
How can I parse this input
```
{
  prop1: 'asd',
  prop2: 'asd'
}
```
 to be 
```
{
  newProp: prop1 || prop2
}
```

So basically how Can I make the input object props transform into new prop and still keep other functions of zodType applicable.
When I do `preprocess` or `transform` I cant then use this schema with other functions like `pick` `omit` `merge` etc

**Comment by JacobWeisenburger** [comment]
This is the best I can come up with based on the limited information you provided.

```ts
z.object( { prop1: z.literal( 'asd' ), prop2: z.literal( 'asd' ) } )
    .transform( ( { prop1, prop2 } ) => ( { newProp: prop1 || prop2 } ) )
```

If you provide more information about your use case, I can try to help you further.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by stychu** [reply]
As I stated 
>When I do preprocess or transform I cant then use this schema with other functions like pick omit merge etc

I can't reuse this schema later on in other schemas If I would like because `.transform` returns ZodEffect instead of ZodType which doesn't allow for further schema composition like merge pick etc

So If I would like to do this I basically can't

```
const SchemaA = z.object( { prop1: z.literal( 'asd' ), prop2: z.literal( 'asd' ) } )
    .transform( ( { prop1, prop2 } ) => ( { newProp: prop1 || prop2 } ) )

const SchemaB = SchemaA.pick({newProp: true}).merge(z.object({abc: z.string()}))   ///<--- Wont work coz of Schema A transfrom
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const SchemaA = z.object( {
    prop1: z.literal( 'foo' ),
    prop2: z.literal( 'bar' ),
} )

const SchemaAWithTransform = SchemaA.transform(
    ( { prop1, prop2 } ) => ( { newProp: prop1 || prop2 } )
)

const SchemaB = z.object( {
    newProp: SchemaA.shape.prop1.or( SchemaA.shape.prop2 ),
} ).merge( z.object( { abc: z.string() } ) )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
